File: golang/day1.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func parseMeasurements(input string) []int {
	lines := strings.Split(input, "\n")

	measurements := make([]int, len(lines))
	for idx, line := range lines {
		measurement, err := strconv.Atoi(line)
		if err != nil {
			fmt.Printf("WARNING: failed to parse measurement (%d)\n", measurement)
		} else {
			measurements[idx] = measurement
		}
	}

	return measurements
}

type IntQueue struct {
	elements chan int
}

func NewIntQueue(size int) IntQueue {
	return IntQueue{
		elements: make(chan int, size),
	}
}

func (queue IntQueue) Enqueue(element int) {
	select {
	case queue.elements <- element:
	default:
		panic("Queue full")
	}
}

func (queue IntQueue) Dequeue() int {
	select {
	case element := <-queue.elements:
		return element
	default:
		panic("Queue empty")
	}
}

func countIncrements(measurements []int, windowSize int) int {

	if windowSize <= 0 {
		message := fmt.Sprintf("Window size (%d) must be positive", windowSize)
		panic(message)
	}

	window := NewIntQueue(windowSize)
	prevWindowSum := 0
	for idx := 0; idx < windowSize; idx++ {
		window.Enqueue(measurements[idx])
		prevWindowSum += measurements[idx]
	}

	increments := 0
	for idx := windowSize; idx < len(measurements); idx++ {
		currWindowSum := prevWindowSum - window.Dequeue() + measurements[idx]

		if prevWindowSum < currWindowSum {
			increments++
		}

		window.Enqueue(measurements[idx])
		prevWindowSum = currWindowSum
	}

	return increments
}

func part1(input string) int {
	measurements := parseMeasurements(input)
	return countIncrements(measurements, 1)
}

func part2(input string) int {
	measurements := parseMeasurements(input)
	return countIncrements(measurements, 3)
}

func main() {
	exitCode := runSolution(part1, part2)
	os.Exit(exitCode)
}

File: python/day1.py
import sys
from collections import deque
from typing import List

from python.util import run_solution


def parse_measurements(input: str) -> List[int]:
    """..."""

    return map(int, input.splitlines())


def count_increments(measurements: List[int], *, window_size: int) -> int:
    """..."""

    assert window_size > 0, f"Window size ({window_size}) must be positive"

    window = deque()
    for index in range(window_size):
        try:
            window.append(next(measurements))
        except StopIteration:
            assert False, (
                f"Window size ({window_size}) must not exceed number of "
                f" measurements ({index + 1})"
            )

    prev_window_sum = sum(window)

    increments = 0
    for curr_measurement in measurements:
        curr_window_sum = prev_window_sum - window.popleft() + curr_measurement
    
        if prev_window_sum < curr_window_sum:
            increments += 1

        window.append(curr_measurement)
        prev_window_sum = curr_window_sum

    return increments


def part1(input: str) -> int:
    """..."""

    measurements = parse_measurements(input)
    return count_increments(measurements, window_size=1)


def part2(input: str) -> int:
    """..."""

    measurements = parse_measurements(input)
    return count_increments(measurements, window_size=3)


if __name__ == "__main__":
    sys.exit(run_solution(part1=part1, part2=part2))